### PR TITLE
fix(collect sequencing times)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

closes: https://github.com/Clinical-Genomics/add-new-tech/issues/62

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/models/run_devices/illumina_run_directory_data.py [hunk 1]
```diff
@@ -97,6 +94,25 @@ def sample_sheet_path(self) -> Path:
             self.get_sequencing_runs_dir(), DemultiplexingDirsAndFiles.SAMPLE_SHEET_FILE_NAME
         )
 
+    @property
+    def get_sequencing_completed_path(self) -> Path:
+        """Return the path to the sequencing completed file."""
+        sequencing_completed_paths: dict = {
```
Comment (ChrOertlin) [inline cg/models/run_devices/illumina_run_directory_data.py:100]: I guess at this point we just have to accept that Illumina changes their way of things over time... anyway hiseq x seem to have a SEQUENCING completed at and NOVASEQ 6000 a SEQUENCE completed at.

File: cg/services/illumina/file_parsing/sequencing_times/hiseq_ga_sequencing_times_service.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+    @staticmethod
+    def get_end_time(run_directory_data: IlluminaRunDirectoryData) -> datetime:
+        """Get the sequencer end date and time."""
+        return run_directory_data.sequenced_at
+
+    @staticmethod
+    def get_start_time(run_directory_data: IlluminaRunDirectoryData) -> datetime:
+        """Get the sequencer start date and time."""
+        return run_directory_data.sequenced_at
```
Comment (ChrOertlin) [inline cg/services/illumina/file_parsing/sequencing_times/hiseq_ga_sequencing_times_service.py:18]: Do we agree on setting both to the same for HISEQGA?

**Review by Vince-janv — APPROVED** [review]
Looks good 👍

File: cg/models/run_devices/illumina_run_directory_data.py (hunk 1 above)
Comment (diitaz93) [reply cg/models/run_devices/illumina_run_directory_data.py:100]: 🤦‍♂️

File: cg/models/run_devices/illumina_run_directory_data.py [hunk 3]
```diff
@@ -97,6 +94,25 @@ def sample_sheet_path(self) -> Path:
...
+            Sequencers.NOVASEQ: Path(
+                self.get_sequencing_runs_dir(),
+                DemultiplexingDirsAndFiles.SEQUENCE_COMPLETED,
+            ),
+            Sequencers.HISEQX: Path(
+                self.get_sequencing_runs_dir(),
+                DemultiplexingDirsAndFiles.SEQUENCING_COMPLETED,
+            ),
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/models/run_devices/illumina_run_directory_data.py:108]: The `get_sequencing_runs_dir` returns the directory of the specific run, so `/home/proj/production/sequencing_data/illumina/sequencing-runs/<flow_cell_id>`.

File: cg/models/run_devices/illumina_run_directory_data.py [hunk 4]
```diff
@@ -97,6 +94,25 @@ def sample_sheet_path(self) -> Path:
...
+        }
+        return (
+            sequencing_completed_paths[self.sequencer_type]
+            if self.sequencer_type in sequencing_completed_paths.keys()
+            else None
+        )
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/models/run_devices/illumina_run_directory_data.py:114]: or
```suggestion
        if self.sequencer_type in sequencing_completed_paths.keys():
            return sequencing_completed_paths[self.sequencer_type]
```

File: cg/models/run_devices/illumina_run_directory_data.py [hunk 5]
```diff
@@ -97,6 +94,25 @@ def sample_sheet_path(self) -> Path:
             self.get_sequencing_runs_dir(), DemultiplexingDirsAndFiles.SAMPLE_SHEET_FILE_NAME
         )
 
+    @property
+    def get_sequencing_completed_path(self) -> Path:
```
Comment (diitaz93) [inline cg/models/run_devices/illumina_run_directory_data.py:98]:
```suggestion
    def get_sequencing_completed_path(self) -> Path | None:
```

File: cg/models/run_devices/illumina_run_directory_data.py [hunk 6]
```diff
@@ -83,7 +80,7 @@ def full_name(self) -> str:
         return self.path.name
 
     @property
-    def sequenced_at(self) -> list[str]:
+    def sequenced_at(self) -> datetime:
         """Return the sequencing date for the sequencing run."""
```
Comment (diitaz93) [inline cg/models/run_devices/illumina_run_directory_data.py:84]:
```suggestion
        """Return the sequencing date from the directory name."""
```

File: cg/models/run_devices/illumina_run_directory_data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/models/run_devices/illumina_run_directory_data.py:108]: Exactly what Sebastian explained.

File: cg/utils/files.py [hunk 7]
```diff
@@ -109,3 +109,14 @@ def get_all_files_in_directory_tree(directory: Path) -> list[Path]:
...
+
+def get_source_modified_time_stamp(source_path: Path) -> float:
+    """
+    Return time stamp that a source is created.
+    Raises:
+        FileNotFoundError if the source does not exist.
+    """
+    if not source_path.exists():
+        raise FileNotFoundError(f"Directory with path {source_path} is not found.")
+    return os.stat(source_path).st_mtime
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/utils/files.py]: source and target are typically used to refer to files or directories I added a docstring.

File: cg/services/illumina/file_parsing/sequencing_times/hiseq_ga_sequencing_times_service.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+"""Module to get the sequencing completed service."""
+
+from datetime import datetime
+from cg.models.run_devices.illumina_run_directory_data import IlluminaRunDirectoryData
+
+
+class HiseqGASequencingTimesService:
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/illumina/file_parsing/sequencing_times/hiseq_ga_sequencing_times_service.py]: I changed it to 2500. although I dont understand why we call it 2500 when the type is  Sequencers.HISEQGA

**Comment by ChrOertlin** [comment]
# test hiseqx
completed succesfully

# test novaseqx 
completed successfully

# test novaseq 6000 
completed successfully

**Pull request merged**